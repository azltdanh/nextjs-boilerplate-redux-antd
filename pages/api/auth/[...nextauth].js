import NextAuth from 'next-auth';
import Providers from 'next-auth/providers';

const options = {
  // pages: {
  //   signIn: '/auth/signin',
  //   signOut: '/auth/signout',
  //   error: '/auth/error', // Error code passed in query string as ?error=
  //   verifyRequest: '/auth/verify-request', // (used for check email message)
  //   newUser: null // If set, new users will be directed here on first sign in
  // },
  providers: [
    // OAuth authentication providers...
    Providers.Facebook({
      clientId: process.env.FACEBOOK_ID,
      clientSecret: process.env.FACEBOOK_SECRET,
    }),
    Providers.Google({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
    }),
    // Password-less / email sign in
    // Providers.Email({
    //   server: process.env.MAIL_SERVER,
    //   from: process.env.MAIL_FROM,
    // }),
  ],

  // Optional SQL or MongoDB database to persist users
  // database: process.env.DATABASE_URL,
};

export default (req, res) => NextAuth(req, res, options);
