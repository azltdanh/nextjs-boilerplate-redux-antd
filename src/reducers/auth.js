import { HYDRATE } from 'next-redux-wrapper';

export const AUTH = {
  // SIGN_IN
  SIGN_IN_REQUEST: 'USER/SIGN_IN_REQUEST',
  SIGN_IN_REQUEST_SUCCESS: 'USER/SIGN_IN_REQUEST_SUCCESS',
  SIGN_IN_REQUEST_FAILURE: 'USER/SIGN_IN_REQUEST_FAILURE',
  // SIGN_OUT
  SIGN_OUT_REQUEST: 'USER/SIGN_OUT_REQUEST',
  SIGN_OUT_REQUEST_SUCCESS: 'USER/SIGN_OUT_REQUEST_SUCCESS',
  SIGN_OUT_REQUEST_FAILURE: 'USER/SIGN_OUT_REQUEST_FAILURE',
  // SIGN_UP
  SIGN_UP_REQUEST: 'USER/SIGN_UP_REQUEST',
  SIGN_UP_REQUEST_SUCCESS: 'USER/SIGN_UP_REQUEST_SUCCESS',
  SIGN_UP_REQUEST_FAILURE: 'USER/SIGN_UP_REQUEST_FAILURE',
};

export const signIn = (payload) => ({
  type: AUTH.SIGN_IN_REQUEST,
  payload,
});

export const signOut = () => ({
  type: AUTH.SIGN_OUT_REQUEST,
});

export const signUp = (payload) => ({
  type: AUTH.SIGN_UP_REQUEST,
  payload,
});

export const initialState = {
  loading: false,
  data: null,
  error: null,
};

const auth = (state = initialState, action = {}) => {
  switch (action.type) {
    case HYDRATE: {
      return {
        ...state,
        ...action.payload.auth, // pick the 'auth' only
      };
    }
    case AUTH.SIGN_IN_REQUEST:
    case AUTH.SIGN_OUT_REQUEST:
    case AUTH.SIGN_UP_REQUEST: {
      return {
        ...state,
        loading: true,
      };
    }
    case AUTH.SIGN_IN_REQUEST_SUCCESS:
    case AUTH.SIGN_UP_REQUEST_SUCCESS: {
      return {
        ...state,
        loading: false,
        data: action.payload,
      };
    }
    case AUTH.SIGN_OUT_REQUEST_SUCCESS: {
      return {
        ...initialState,
      };
    }
    case AUTH.SIGN_IN_REQUEST_FAILURE:
    case AUTH.SIGN_OUT_REQUEST_FAILURE:
    case AUTH.SIGN_UP_REQUEST_FAILURE: {
      return {
        ...state,
        loading: false,
        error: action.payload,
      };
    }
    default: {
      return { ...state };
    }
  }
};

export default auth;
