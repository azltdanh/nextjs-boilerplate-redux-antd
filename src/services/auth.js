import api from './api';

export const signIn = async ({ email, password }) => {
  const resp = await api.get(`/users?email=${email}&password=${password}`);
  return resp.data[0];
};

export const signOut = async (payload) => {
  const resp = await api.post('/sign-out', payload).catch(() => ({})); // mock api
  return resp.data;
};

export const signUp = async (payload) => {
  const resp = await api.post('/users', payload);
  return resp.data;
};
