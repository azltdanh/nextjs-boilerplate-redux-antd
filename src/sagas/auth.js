import { all, call, put, takeLatest } from 'redux-saga/effects';
import { signIn, signOut, signUp } from '@services/auth';
import { AUTH } from '@reducers/auth';

export function* signInSaga(action) {
  try {
    const data = yield call(signIn, action.payload);
    yield put({ type: AUTH.SIGN_IN_REQUEST_SUCCESS, payload: data });
  } catch (error) {
    yield put({ type: AUTH.SIGN_IN_REQUEST_FAILURE, payload: error });
  }
}

export function* signOutSaga(action) {
  try {
    const data = yield call(signOut, action.payload);
    yield put({ type: AUTH.SIGN_OUT_REQUEST_SUCCESS, payload: data });
  } catch (error) {
    yield put({ type: AUTH.SIGN_OUT_REQUEST_FAILURE, payload: error });
  }
}

export function* signUpSaga(action) {
  try {
    const data = yield call(signUp, action.payload);
    yield put({ type: AUTH.SIGN_UP_REQUEST_SUCCESS, payload: data });
  } catch (error) {
    yield put({ type: AUTH.SIGN_UP_REQUEST_FAILURE, payload: error });
  }
}

function* authSagas() {
  yield all([
    takeLatest(AUTH.SIGN_IN_REQUEST, signInSaga),
    takeLatest(AUTH.SIGN_OUT_REQUEST, signOutSaga),
    takeLatest(AUTH.SIGN_UP_REQUEST, signUpSaga),
  ]);
}

export default authSagas;
