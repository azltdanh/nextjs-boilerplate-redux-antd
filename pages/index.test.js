import { useSelector } from 'react-redux';
import * as authReducers from '@reducers/auth';
import mockUser from '@mock/user';
import Home from '.';

// mock router history
const mockHistoryPush = jest.fn();
jest.mock('next/router', () => ({
  ...jest.requireActual('next/router'),
  useRouter: () => ({
    pathname: '/',
    push: mockHistoryPush,
  }),
}));

// mock redux store
const mockUseSelector = (mockedStore) => {
  useSelector.mockImplementation((callback) => callback(mockedStore));
};
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

describe(Home.name, () => {
  beforeEach(() => {
    mockDispatch.mockClear();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders', () => {
    const mockStore = {
      auth: { ...authReducers.initialState, data: { ...mockUser } },
    };
    mockUseSelector(mockStore);
    expect(shallow(<Home />)).toMatchSnapshot();
  });
});
