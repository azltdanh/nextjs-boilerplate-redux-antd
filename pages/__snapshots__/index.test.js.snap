// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders 1`] = `
<section>
  <h2>
    Home
  </h2>
  <h3>
    1.User 1
  </h3>
</section>
`;
